numaproj/numaplane
# fix:  Numaflow Controller deletion - use gitops-engine settings to delete/recreate instead of doing it ourselves

Fixes #455 #377 

### Modifications

This is a follow-up to [this](https://github.com/numaproj/numaplane/pull/464) previous PR. That PR made it so that if the version number changes, then the previous Numaflow Controller manifests are deleted and then recreated. However, it doesn't take care of the cases in which the version number doesn't change but there's some other change to the manifest of a given version. (In fact, we saw this issue when deploying.)

I noticed that gitops-engine can actually be configured to do a "Replace" with force rather than an "Apply", which effectively handles the problem of backward compatibility rather than our needing to explicitly delete the manifest first. 

Note that it is also able to detect whether there is a difference before doing anything: in the log I can see that when there is no difference in a given resource, it just says "Skipping as resource was not modified".

This PR also makes a few other minor fixes which are described inline.


### Verification

#### Updated e2e test

The e2e test now changes the Numaflow Controller manifest from one version to another in a way which previously would've failed, due to an incompatible change.
In addition, the e2e test verifies the change of the NumaflowControllerRolllout to an invalid version and back again.

#### Manual testing

Performed the following test (multiple times to ensure consistency) to test the path we plan to take in Production going from Numaplane 0.9.1:

1. Started from 0.9.1 tag and updated the 1.3.3 manifest to remove "InstanceID" field from all of the labels
2. Created NumaflowControllerRollout at 1.3.3, ISBServiceRollout, and PipelineRollout
3. Then switched to this branch and started Numaplane
4. Observed everything still running healthily - NumaflowController CR was created since that's new
5. Updated NumaflowControllerRollout to 1.4.0 manifest which contains the "InstanceID" label. Observed Pipeline pause, manifest update to 1.4.0, pipelines unpause. Observed NumaflowControllerRollout "Deployed" with child condition healthy
6. Updated NumaflowControllerRollout to 99.99.99, which doesn't exist. Observed Pipeline pause, manifest can't be found, so NumaflowController marked "Failed" and NumaflowControllerRollout HealthyChildCondition marked False ("degraded"). Numaflow Controller pod still running 1.4.0 and pipeline unpaused.
7. Updated NumaflowControllerRollout back to 1.4.0 and observed NumaflowController went back to "Deployed" and NumaflowControllerRollout HealthyChildCondition back to True.



### Backward incompatibilities

I have explicitly tested the path from 0.9.1 as described above. I don't think there are any issues.
